indra/newview/llappearancelistener.cpp

```diff
@@ -0,0 +1,165 @@
+        &LLAppearanceListener::wearItem,
+        llsd::map("item_id", LLSD(), "replace", LLSD()));
+
+    add("detachItem",
+        "Detach item by item id: [item_id]",
+        &LLAppearanceListener::detachItem,
+        llsd::map("item_id", LLSD()));
+
+    add("getOutfitsList",
+        "Return the table with Outfits info(id and name)",
+         &LLAppearanceListener::getOutfitsList);
+
+    add("getOutfitItems",
+        "Return the table of items with info(id : name, wearable_type, is_worn) inside specified outfit folder",
+         &LLAppearanceListener::getOutfitItems);
+}
+
+
+void LLAppearanceListener::wearOutfit(LLSD const &data)
+{
+    Response response(LLSD(), data);
+    LLViewerInventoryCategory* cat = gInventory.getCategory(data["folder_id"].asUUID());
+    if (!cat)
+    {
+        response.error(stringize(LLTrans::getString("OutfitNotFound"), data["folder_id"].asUUID()));
+        return;
+    }
+    if (LLFolderType::lookupIsProtectedType(cat->getPreferredType()))
+    {
+        response.error(stringize(LLTrans::getString("SystemFolderNotWorn"), data["folder_id"].asUUID()));
+        return;
+    }
+    bool append = data["append"].asBoolean();
+    bool can_wear = append ? LLAppearanceMgr::instance().getCanAddToCOF(cat->getUUID()) : LLAppearanceMgr::instance().getCanReplaceCOF(cat->getUUID());
+    if (!can_wear)
+    {
+        std::string msg = append ? LLTrans::getString("OutfitNotAdded") : LLTrans::getString("OutfitNotReplaced");
+        response.error(stringize(msg, std::quoted(cat->getName()), " , id: ", cat->getUUID()));
+        return;
+    }
+    LLAppearanceMgr::instance().wearInventoryCategory(cat, false, append);
+}
+
+void LLAppearanceListener::wearOutfitByName(LLSD const &data)
+{
+    Response response(LLSD(), data);
+    std::string error_msg;
+    if (!LLAppearanceMgr::instance().wearOutfitByName(data["folder_name"].asString(), error_msg, data["append"].asBoolean()))
+    {
+        response.error(error_msg);
+    }
+}
+
+void LLAppearanceListener::wearItem(LLSD const &data)
+{
+    LLAppearanceMgr::instance().wearItemOnAvatar(data["item_id"].asUUID(), true, data["replace"].asBoolean());
+}
+
+void LLAppearanceListener::detachItem(LLSD const &data)
+{
+    LLAppearanceMgr::instance().removeItemFromAvatar(data["item_id"].asUUID());
+}
+
+void LLAppearanceListener::getOutfitsList(LLSD const &data)
+{
+    Response response(LLSD(), data);
+    const LLUUID outfits_id = gInventory.findCategoryUUIDForType(LLFolderType::FT_MY_OUTFITS);
+
+    LLInventoryModel::cat_array_t cat_array;
+    LLInventoryModel::item_array_t item_array;
+
+    LLIsType is_category(LLAssetType::AT_CATEGORY);
+    gInventory.collectDescendentsIf(outfits_id, cat_array, item_array, LLInventoryModel::EXCLUDE_TRASH, is_category);
+
+    LLSD outfits_data;
+    for (const LLPointer<LLViewerInventoryCategory> &cat : cat_array)
+    {
+        outfits_data[cat->getUUID().asString()] = cat->getName();
+    }
```

Comment:
You inspired me. I looked at these loops and thought: this is a pattern I've seen over and over again, how about [a utility function](https://github.com/secondlife/viewer/pull/1924)?

I stipulate that whether it's clearer in this case is a matter of style. Feel free to tell me that the functions are interesting but the explicit loop makes more sense.

Reply:
Actually, it looks good in `getOutfitItems` where there are multiple info fields in the response.